fsspec/implementations/cached.py

```diff
@@ -255,11 +253,11 @@ def clear_expired_cache(self, expiry_time=None):
 
         for path, detail in self.cached_files[-1].copy().items():
             if time.time() - detail["time"] > expiry_time:
-                if self.same_names:
-                    basename = os.path.basename(detail["original"])
-                    fn = os.path.join(self.storage[-1], basename)
-                else:
-                    fn = os.path.join(self.storage[-1], detail["fn"])
+                fn = getattr(detail, "fn", "")
+                if not fn:
+                    # fn should always be set, but be defensive here.
```

Comment:
Feel like if we have a condition we expect never to hit, it's probably an exception or warning. I suppose this line is not covered by tests?

Reply:
I've changed this to raise an exception, and added an explicit test that the exception is raised in `test_cached.py::test_clear_expired` as that already contains the machinery to modify the cached metadata.